File: AISuite_QuickStart/app/src/main/java/com/zebra/aisuite_quickstart/java/detectors/productrecognition/ProductRecognitionAnalyzer.java
// Copyright 2025 Zebra Technologies Corporation and/or its affiliates. All rights reserved.
package com.zebra.aisuite_quickstart.java.detectors.productrecognition;

import android.util.Log;

import androidx.annotation.NonNull;
import androidx.camera.core.ImageAnalysis;
import androidx.camera.core.ImageProxy;

import com.zebra.ai.vision.detector.AIVisionSDKException;
import com.zebra.ai.vision.detector.BBox;
import com.zebra.ai.vision.detector.ImageData;
import com.zebra.ai.vision.detector.ModuleRecognizer;
import com.zebra.ai.vision.entity.Entity;
import com.zebra.ai.vision.entity.ShelfEntity;

import java.util.List;
import java.util.Objects;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;


/**
 * The ProductRecognitionAnalyzer class implements the ImageAnalysis.Analyzer interface and is
 * responsible for analyzing image frames to detect and recognize products. It integrates module recognizer for matching and
 * identify products.
 *
 * This class is designed to be used within an Android application as part of a camera-based
 * product recognition solution. It processes image data asynchronously and returns recognition
 * results through a callback interface.
 *
 * Usage:
 * - Instantiate the ProductRecognitionAnalyzer with the appropriate callback, localizer, feature extractor, and recognizer.
 * - Implement the DetectionCallback interface to handle recognition results.
 * - The analyze(ImageProxy) method is called by the camera framework to process image frames.
 * - Call stopAnalyzing() to stop the analysis process and release resources.
 *
 * Dependencies:
 * - Android ImageProxy: Provides access to image data from the camera.
 * - ExecutorService: Used for asynchronous task execution.
 * - ModuleRecognizer
 *
 * Concurrency:
 * - Uses a single-threaded executor to ensure that image analysis tasks are processed sequentially.
 * - Manages concurrency with flags to control analysis state and termination.
 *
 * Note: Ensure that the appropriate permissions and dependencies are configured
 * in the AndroidManifest and build files to utilize camera and image processing capabilities.
 */
public class ProductRecognitionAnalyzer implements ImageAnalysis.Analyzer {

    /**
     * Interface for handling the results of the product recognition process.
     * Implement this interface to define how recognition results are processed.
     */
    public interface DetectionCallback {
        void onRecognitionResult(List<Entity> result);
    }

    private static final String TAG = "ProductRecognitionAnalyzer";
    private boolean isAnalyzing = true;
    private final DetectionCallback callback;
    private BBox[] detections, products;
    private volatile boolean isStopped = false;
    private final ExecutorService executorService;
    private final ModuleRecognizer productRecognizer;


    public ProductRecognitionAnalyzer(DetectionCallback callback, ModuleRecognizer productRecognizer) {
        this.callback = callback;
        this.productRecognizer = productRecognizer;
        this.executorService = Executors.newSingleThreadExecutor();
    }

    /**
     * Analyzes the given image to perform product recognition. This method is called by the camera
     * framework to process image frames asynchronously.
     *
     * @param image The image frame to analyze.
     */
    @Override
    public void analyze(@NonNull ImageProxy image) {
        Log.e(TAG, "analyze() called");
        if(productRecognizer == null){
            Log.d(TAG, "Module recognizer is null");
            image.close();
            return;
        }
        if (!isAnalyzing || isStopped) {
            Log.d(TAG, "Analyzer is stopped or already analyzing. Closing image.");
            image.close();
            return;
        }

        isAnalyzing = false;

        Log.d(TAG, "Converting ImageProxy to Bitmap...");
        ImageData imageData = ImageData.fromImageProxy(image);

        Log.d(TAG, "Calling moduleRecognizer.process...");
        long start = System.currentTimeMillis();
        try {
            productRecognizer.process(imageData)
                    .thenAccept(entityList -> {
                        long end = System.currentTimeMillis();
                        long inferenceTime = end - start;
                        Log.d(TAG, "Inference Time: " + inferenceTime);
                        int shelfCount = 0;
                        if (entityList != null) {
                            for (Entity entity : entityList) {
                                if (entity instanceof ShelfEntity) {
                                    shelfCount++;
                                }
                            }
                        }
                        Log.d(TAG, "process() completed. Shelves found: " + shelfCount);
                        if (!isStopped && callback != null) {
                            Log.d(TAG, "Invoking callback.onRecognitionResult");
                            callback.onRecognitionResult(entityList);
                        }
                        Log.d(TAG, "Image closed, ready for next frame.");
                    })
                    .exceptionally(ex -> {
                        Log.e(TAG, "Error in shelf recognition: " + ex.getMessage(), ex);
                        return null;
                    });
            image.close();
            isAnalyzing = true;
        } catch (AIVisionSDKException e) {
            Log.e(TAG, Objects.requireNonNull(e.getMessage()));
            isAnalyzing = true;
            image.close();
        }
    }


    /**
     * Stops the analysis process and terminates any ongoing tasks. This method should be
     * called to release resources and halt image analysis when it is no longer required.
     */

    public void stopAnalyzing() {
        Log.d(TAG, "stopAnalyzing() called. Shutting down executor.");
        isStopped = true;
        executorService.shutdownNow();
    }
}

File: AISuite_QuickStart/app/src/main/java/com/zebra/aisuite_quickstart/kotlin/detectors/productrecognition/ProductRecognitionAnalyzer.kt
// Copyright 2025 Zebra Technologies Corporation and/or its affiliates. All rights reserved.
package com.zebra.aisuite_quickstart.kotlin.detectors.productrecognition

import android.util.Log
import androidx.camera.core.ImageAnalysis
import androidx.camera.core.ImageProxy
import com.zebra.ai.vision.detector.ImageData
import com.zebra.ai.vision.detector.ModuleRecognizer
import com.zebra.ai.vision.entity.Entity
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.launch
import kotlinx.coroutines.suspendCancellableCoroutine
import kotlinx.coroutines.withContext
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException

class ProductRecognitionAnalyzer(
    private val callback: DetectionCallback,
    private val moduleRecognizer: ModuleRecognizer?
) : ImageAnalysis.Analyzer {

    interface DetectionCallback {
        fun onRecognitionResult(result: List<Entity>?)
        fun onCaptureRecognitionResult(result: List<Entity>?)
    }

    private val TAG = "ProductRecognitionAnalyzer"
    private var isAnalyzing = true
    private var isStopped = false
    private var job = Job()
    private var scope = CoroutineScope(Dispatchers.IO + job)

    override fun analyze(image: ImageProxy) {
        if (moduleRecognizer == null) {
            Log.d(TAG, "moduleRecognizer is null")
            image.close()
            return
        }
        if (!isAnalyzing || isStopped) {
            image.close()
            return
        }
        isAnalyzing = false

        scope.launch {
            try {
                val imageData = ImageData.fromImageProxy(image)
                val start = System.currentTimeMillis()
                moduleRecognizer.process(imageData).thenAccept { entityList ->
                    val end = System.currentTimeMillis()
                    Log.d(TAG, "Inference Time: ${end - start} ms")
                    if (!isStopped) {
                        callback.onRecognitionResult(entityList)
                    }
                    image.close()
                    isAnalyzing = true
                }.exceptionally { ex ->
                    Log.e(TAG, "Error in product recognition: ${ex.message}", ex)
                    image.close()
                    isAnalyzing = true
                    null
                }
            } catch (ex: Exception) {
                Log.e(TAG, "Exception during analyze: ${ex.message}", ex)
                image.close()
                isAnalyzing = true
            }
        }
    }

    /**
     * Processes a captured image using the provided capture recognizer and delivers
     * results via the onCaptureRecognitionResult callback.
     *
     * @param image The ImageProxy containing the captured image data to process.
     * @param captureRecognizer The ModuleRecognizer instance to use for capture processing.
     */
    fun processImage(image: ImageProxy, captureRecognizer: ModuleRecognizer) {
        val captureScope = CoroutineScope(Dispatchers.IO + Job())
        captureScope.launch {
            try {
                Log.d(TAG, "Starting image capture recognition analysis")
                val result = suspendCancellableCoroutine<List<Entity>> { cont ->
                    try {
                        val imageData = ImageData.fromImageProxy(image)
                        captureRecognizer.process(imageData)
                            .thenAccept { entityList ->
                                cont.resume(entityList)
                            }
                            .exceptionally { ex ->
                                cont.resumeWithException(ex)
                                null
                            }
                    } catch (e: Exception) {
                        cont.resumeWithException(e)
                    }
                }
                withContext(Dispatchers.Main) {
                    callback.onCaptureRecognitionResult(result)
                }
            } catch (ex: Exception) {
                Log.e(TAG, "Error in capture recognition processing: ${ex.message}", ex)
            } finally {
                image.close()
                isAnalyzing = true
            }
        }
    }

    /**
     * Stops the analysis process and cancels the coroutine job.
     */
    fun stopAnalyzing() {
        isStopped = true
        job.cancel()
    }

    /**
     * Starts or restarts the analysis process. This method resets the stopped state
     * and creates a new coroutine scope for processing.
     */
    fun startAnalyzing() {
        Log.d(TAG, "startAnalyzing() called.")
        isStopped = false
        job = Job()
        scope = CoroutineScope(Dispatchers.IO + job)
    }

}